File: src/horizon_mcp/tools/reports.py
"""Report management tools for Horizon MCP Server.

3 tools covering report listing, CSV download, and deletion:
  - list_reports: list reports with optional name filter and expiry toggle
  - download_report: fetch raw CSV content by report UUID
  - delete_report: delete a report by UUID with safety echo

CRITICAL path note:
  - CSV downloads use ``/reports/{uuid}`` (NO ``/api/v1`` prefix).
  - API management (list / delete) uses ``/api/v1/reports/``.
"""

from __future__ import annotations

import json
import logging
from typing import TYPE_CHECKING, Any

from horizon_mcp.client.errors import HorizonError
from horizon_mcp.client.state import get_client
from horizon_mcp.tools._helpers import delete_guard

if TYPE_CHECKING:
    from mcp.server.fastmcp import FastMCP

logger = logging.getLogger("horizon_mcp.tools.reports")

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

_REPORT_API_BASE = "/api/v1/reports"
_REPORT_CSV_BASE = "/reports"


# ---------------------------------------------------------------------------
# Registration
# ---------------------------------------------------------------------------

def register_report_tools(mcp: FastMCP) -> None:
    """Register all 3 report tools on *mcp*."""

    @mcp.tool()
    async def list_reports(
        max_items: int = 50,
        report_name: str | None = None,
        expired: bool = False,
    ) -> str:
        """List available reports, optionally filtered by name.

        Safety tier: read-only

        When *report_name* is provided the server returns all report entries
        matching that name (there can be more than one).  Without a name the
        full report catalogue is returned.

        Args:
            max_items: Maximum items to return (default 50).
            report_name: Exact report name to filter on (server-side).
            expired: Include expired reports (default false).

        Returns:
            JSON with items, count, total_available, truncated flag, and kind.
        """
        client = get_client()
        params: dict[str, str] = {"expired": str(expired).lower()}
        path = (
            f"{_REPORT_API_BASE}/{report_name}"
            if report_name
            else _REPORT_API_BASE
        )
        data = await client.get(path, params=params)

        items: list[dict[str, Any]] = (
            data if isinstance(data, list) else [data] if data else []
        )
        total = len(items)
        truncated = total > max_items
        items = items[:max_items]

        return json.dumps({
            "items": items,
            "count": len(items),
            "total_available": total,
            "truncated": truncated,
            "kind": "report",
        })

    @mcp.tool()
    async def download_report(report_uuid: str) -> str:
        """Download a report as CSV by its UUID.

        Safety tier: read-only

        CRITICAL: The CSV endpoint lives at ``/reports/{uuid}`` — there is
        NO ``/api/v1`` prefix for this path.

        Args:
            report_uuid: UUID of the report to download.

        Returns:
            JSON with a summary message, the raw CSV text, and a row count.
        """
        client = get_client()
        resp = await client._request("GET", f"{_REPORT_CSV_BASE}/{report_uuid}")
        csv_text = resp.text

        lines = csv_text.strip().splitlines()
        row_count = max(0, len(lines) - 1) if lines else 0

        return json.dumps({
            "content": f"Report {report_uuid} downloaded ({row_count} rows).",
            "csv": csv_text,
            "rows": row_count,
        })

    @mcp.tool()
    async def delete_report(report_uuid: str, expected_uuid: str) -> str:
        """Delete a report by UUID. Requires UUID confirmation.

        Safety tier: mutating-destructive

        IMPORTANT: Before executing this operation, always confirm the
        action with the end-user first.

        Args:
            report_uuid: UUID of the report to delete.
            expected_uuid: Must exactly match *report_uuid* as a deletion
                safeguard.

        Returns:
            JSON confirmation of deletion.
        """
        delete_guard(report_uuid, expected_uuid, label="uuid")
        client = get_client()
        await client.delete(f"{_REPORT_API_BASE}/{report_uuid}")
        return json.dumps({
            "deleted": True,
            "uuid": report_uuid,
            "kind": "report",
        })

File: src/tools/reports.ts
/**
 * Report management tools for Horizon MCP Server.
 *
 * 3 tools covering report listing, CSV download, and deletion:
 *   - list_reports: list reports with optional name filter and expiry toggle
 *   - download_report: fetch raw CSV content by report UUID
 *   - delete_report: delete a report by UUID with safety echo
 *
 * CRITICAL path note:
 *   - CSV downloads use /reports/{uuid} (NO /api/v1 prefix).
 *   - API management (list / delete) uses /api/v1/reports/.
 */
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';

import type { HorizonClient } from '../client/http.js';
import { deleteGuard } from './helpers.js';
import { registerTool } from './register.js';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const REPORT_API_BASE = '/api/v1/reports';
const REPORT_CSV_BASE = '/reports';

// ---------------------------------------------------------------------------
// Registration
// ---------------------------------------------------------------------------

export function registerReportTools(
  server: McpServer,
  client: HorizonClient,
): void {
  registerTool(
    server,
    'list_reports',
    {
      description:
        'List available reports, optionally filtered by name.\n\n' +
        'Safety tier: read-only\n\n' +
        'When report_name is provided the server returns all report entries ' +
        'matching that name (there can be more than one). Without a name the ' +
        'full report catalogue is returned.',
      inputSchema: z.object({
        max_items: z
          .number()
          .int()
          .positive()
          .max(100)
          .default(50)
          .describe('Maximum items to return (default 50).'),
        report_name: z
          .string()
          .optional()
          .describe('Exact report name to filter on (server-side).'),
        expired: z
          .boolean()
          .default(false)
          .describe('Include expired reports (default false).'),
      }),
    },
    async ({ max_items, report_name, expired }) => {
      const params = new URLSearchParams({
        expired: String(expired),
      });
      const path = report_name
        ? `${REPORT_API_BASE}/${report_name}`
        : REPORT_API_BASE;

      const data = await client.get<unknown>(path, params);
      const items: Record<string, unknown>[] = Array.isArray(data)
        ? (data as Record<string, unknown>[])
        : data
          ? [data as Record<string, unknown>]
          : [];

      const total = items.length;
      const truncated = total > max_items;
      const sliced = items.slice(0, max_items);

      return {
        content: [
          {
            type: 'text' as const,
            text: JSON.stringify({
              items: sliced,
              count: sliced.length,
              total_available: total,
              truncated,
              kind: 'report',
            }),
          },
        ],
      };
    },
  );

  registerTool(
    server,
    'download_report',
    {
      description:
        'Download a report as CSV by its UUID.\n\n' +
        'Safety tier: read-only\n\n' +
        'CRITICAL: The CSV endpoint lives at /reports/{uuid} - there is ' +
        'NO /api/v1 prefix for this path.',
      inputSchema: z.object({
        report_uuid: z.string().describe('UUID of the report to download.'),
      }),
    },
    async ({ report_uuid }) => {
      const csvText = await client.getText(`${REPORT_CSV_BASE}/${report_uuid}`);

      const lines = csvText.trim().split('\n');
      const rowCount = lines.length > 0 ? Math.max(0, lines.length - 1) : 0;

      return {
        content: [
          {
            type: 'text' as const,
            text: JSON.stringify({
              content: `Report ${report_uuid} downloaded (${rowCount} rows).`,
              csv: csvText,
              rows: rowCount,
            }),
          },
        ],
      };
    },
  );

  registerTool(
    server,
    'delete_report',
    {
      description:
        'STOP - This tool performs an IRREVERSIBLE destructive operation. You MUST ' +
        'ask the user for explicit confirmation before calling this tool. Do not ' +
        'proceed without a clear "yes" from the user. Present what will be ' +
        'permanently destroyed and wait.\n\n' +
        'Delete a report by UUID. Requires UUID confirmation.\n\n' +
        'Safety tier: mutating-destructive',
      inputSchema: z.object({
        report_uuid: z.string().describe('UUID of the report to delete.'),
        expected_uuid: z
          .string()
          .describe('Must exactly match report_uuid as a deletion safeguard.'),
      }),
    },
    async ({ report_uuid, expected_uuid }) => {
      deleteGuard(report_uuid, expected_uuid, 'uuid');
      await client.delete(`${REPORT_API_BASE}/${report_uuid}`);
      return {
        content: [
          {
            type: 'text' as const,
            text: JSON.stringify({
              deleted: true,
              uuid: report_uuid,
              kind: 'report',
            }),
          },
        ],
      };
    },
  );
}
